File: without-library/basics/lis_lcs.py
import sys

# LIS / LCS without libraries

def lis(arr):
    if not arr: return 0
    n = len(arr)
    dp = [1] * n
    
    for i in range(1, n):
        for j in range(i):
            if arr[i] > arr[j]:
                if dp[j] + 1 > dp[i]:
                    dp[i] = dp[j] + 1
                    
    # Manual max
    mx = 0
    for x in dp:
        if x > mx: mx = x
    return mx

def lcs(s1, s2):
    n = len(s1)
    m = len(s2)
    dp = [[0] * (m + 1) for _ in range(n + 1)]
    
    for i in range(1, n + 1):
        for j in range(1, m + 1):
            if s1[i-1] == s2[j-1]:
                dp[i][j] = dp[i-1][j-1] + 1
            else:
                v1 = dp[i-1][j]
                v2 = dp[i][j-1]
                dp[i][j] = v1 if v1 > v2 else v2
                
    return dp[n][m]

if __name__ == "__main__":
    pass

File: without-library/basics/lis_lcs.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

int lis(vector<int>& arr) {
    if (arr.empty()) return 0;
    int n = arr.size();
    vector<int> dp(n, 1);
    
    for (int i = 1; i < n; i++) {
        for (int j = 0; j < i; j++) {
            if (arr[i] > arr[j]) {
                if (dp[j] + 1 > dp[i]) {
                    dp[i] = dp[j] + 1;
                }
            }
        }
    }
    
    int mx = 0;
    for (int x : dp) if (x > mx) mx = x;
    return mx;
}

int lcs(string s1, string s2) {
    int n = s1.length();
    int m = s2.length();
    vector<vector<int>> dp(n + 1, vector<int>(m + 1, 0));
    
    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= m; j++) {
            if (s1[i-1] == s2[j-1]) {
                dp[i][j] = dp[i-1][j-1] + 1;
            } else {
                int v1 = dp[i-1][j];
                int v2 = dp[i][j-1];
                dp[i][j] = (v1 > v2) ? v1 : v2;
            }
        }
    }
    
    return dp[n][m];
}

int main() {
    return 0;
}
